lib/util/pre_receive_util.js

```diff
@@ -0,0 +1,315 @@
+const checkSubmodule = Symbol();
+const checkTree = Symbol();
+const parentLoop = Symbol();
+class PreReceiveCheck {
+
+    /**
+     * (This has to be public so we can mock it for testing)
+     * @param {NodeGit.Commit} commit
+     */
+    getSyntheticBranchForCommit(commit) {
+        return "refs/commits/" + commit.id();
+    }
+
+    constructor() {
+        this[urlToLocalPath] = function*(repo, url) {
+            const config = yield repo.config();
+            const subrepoUrlBase =
+                yield config.getStringBuf("gitmeta.subrepourlbase");
+            const subrepoRootPath =
+                yield config.getStringBuf("gitmeta.subreporootpath");
+            assert(url.startsWith(subrepoUrlBase));
+            const remotePath = url.slice(subrepoUrlBase.length);
+            const localPath = path.join(subrepoRootPath, remotePath);
+            if (localPath[0] === "/") {
+                return localPath;
+            } else {
+                return path.normalize(path.join(repo.path(), localPath));
+            }
+        };
+
+        /**
+         * Check that a synthetic metabranch exists for a given submodule
+         * at a given commit.
+         * @async
+         * @param {NodeGit.Repostory} repo The meta repository
+         * @param {NodeGit.Commit} commit The meta branch's most recent commit
+         * @param {NodeGit.TreeEntry} submoduleEntry the submodule's tree entry
+         * in the meta tree.
+         */
+        this[checkSubmodule] = function *(repo, commit, submoduleEntry) {
+            const submodulePath = submoduleEntry.path();
+            const submodule = yield NodeGit.Submodule.lookup(repo,
+                                                             submodulePath);
+            const url = submodule.url();
+            const localPath = yield *this[urlToLocalPath](repo, url);
+            const submoduleRepo = yield NodeGit.Repository.open(localPath);
+            const submoduleCommitId = submoduleEntry.id();
+            const branch = this.getSyntheticBranchForCommit(submoduleCommitId);
+            try {
+                const subrepoCommit =
+                    yield submoduleRepo.getReferenceCommit(branch);
+                return subrepoCommit.id().equal(submoduleEntry.id());
+            } catch (e) {
+                return false;
+            }
+        };
+
+        /**
+         * Check a tree.  For each entry, if the entry is identical to its
+         * parent, it's OK.  Blobs are always OK.  And submodules are
+         * checked with checkSubmodule.
+         * @async
+         * @param {NodeGit.Repostory} repo The meta repository
+         * @param {NodeGit.Commit} commit The meta branch's commit to check
+         * @param {NodeGit.Tree[]} parentTrees the tree objects in parent
+         * commits that correspond to this tree
+         * @param {NodeGit.Tree} tree The tree to check
+         */
+        this[checkTree] = function *(repo, commit, parentTrees, tree) {
```

Comment:
I believe libgit2 provides a way to look at what's changed in a commit, e.g. `NodeGit.Commit.getDiffWithOptions` that might be simpler (you won't need this recursive tree processing logic) and more efficient. Also, since we really only care about submodules you could probably use `SubmoduleUtil.getSubmoduleChanges`, which will tell you want submodules have been added/changed/removed in a commit. It looks like its documentation is incorrect in that it returns sets, not maps; I'll change.